File: trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, hi: List[int]) -> int:
        water_units = 0
        for i in range(0, len(hi)):
            left = hi[:i]
            right = hi[i+1:]
            min_max = min(max(left, default=0), max(right, default=0))
            if min_max > hi[i]:
                water_units += min_max - hi[i]
        return water_units

File: trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        if (height.length == 0) {
            return 0;
        }
        int[] leftMax = new int[height.length], rightMax = new int[height.length];
        leftMax[0] = 0;
        rightMax[height.length-1] = 0;
        
        for(int i=1; i<height.length; i++) {
            leftMax[i] = Math.max(leftMax[i-1], height[i-1]);
        }
        
        for(int j=height.length-2; j>=0; j--) {
            rightMax[j] = Math.max(rightMax[j+1], height[j+1]);
        }
        
        int waterTrapped = 0;
        
        for(int i=0; i<height.length; i++) {
            waterTrapped += Math.max(Math.min(leftMax[i], rightMax[i]) - height[i], 0);
        }
        
        return waterTrapped;
    }
}